### DISCUSSION TITLE
Whats next?

### DISCUSSION THREAD

**Discussion by snowzach** [iamromulan/cellular-modem-wiki#61] [Ideas/Requests - Quectel AP Repo]
What's next?

Anyone else with a RM521 have regular disconnects? I never use 3CA and was wondering if I could get someone to trade a RM521 for a RM520 if it's more stable.

Did we every figure out an auto-reboot script if it looses connection?

**Comment by iamromulan** [comment]
@snowzach 
I believe the issue is infrequent on the 520 compared to the 521 but it still happens.
I have identified where the PCIe Ethernet chipset disappears as seen with dmesg here: 
https://forums.quectel.com/t/rm5xx-m-2-pci-e-ethernet-stops-responding/33178

This was running on T-Mobile.
AT&T seems this problem is almost non-existent

Watch: development: /simpleadmin/console/services

Creating ping_watchdog and Ethernet_watchdog

It will be configurable through the console menu that will launch when you login to console. 

I was also thinking about having 80 serve content from simpleadmin/http_www/ 
Only basic stats and a login button that redirects to 443; what do you think?

**Comment by snowzach** [comment]
Hmmm... So I was actually thinking about writing a Go server that could replace:
* lighttpd
* atcmd
* Could easily be a configurable watchdog

An all in one might be kind of nice... It could also remove the need for opkg (although we may want that anyhow)

I was trying to figure out how the socat-at-bridge worked.. It's kinda confusing using socat and stitching like 4 ttys together. 
Anyhow, it might be kinda nice to have a simple interface where you could curl/wget http(s)://localhost/atcmd?=AT+WHATEVER and it would return the response... It could handle contention and throttling really easily...

**Reply by iamromulan** [reply]
@snowzach 
There are 3 separate internal dev that are used for AT communications (3 separate AT command streams) 

at_mdm0
smd7
smd11

Socat creates normal tty devices for our use.

at_mdm0 and SMD7 are used by making them available via USB with the port_bridge.service
smd11 is available/unused
I recently gained access to a bunch of documentation on this (where as before we didn't have any documentation at all and we had to figure this out on our own)

I will be uploading all these docs soon. 😁

If you want to give Go....a go lol go for it 😂

Unfortunately though the /usrdata partition (ubi2) only has less than 100mb to work with and we only have a little single core 32bit armv7 to work with. That's why I stuck with bash scripts and compiled binaries for the most part. Micropython ended up being too much for it, maybe Go would do well. Not too sure. I can tell you that the new modules on the way RM550 and RM551 are supposed to have quad core arm64 (v8) onboard and a bit more space. They are even supposed to have an OpenWRT implementation on them as well.

![IMG20240430104446.jpg](https://github.com/iamromulan/quectel-rgmii-toolkit/assets/50184035/0ae83981-85fb-418c-9564-a2727342c287)

**Reply by snowzach** [reply]
Interesting! Yeah, I spent a good amount of time trying to open the smd7/smd11 device in go last night.. It just doesn't work... But somehow piping it through socat make it's happy.. I was trying to figure out what cat+socat did to make it work. Why can cat read them but a go program can't. Kept getting `device or resource is busy` (after verifying nothing else was using it after killing cat/socat/etc)

The idea of using Go would be a single executable that could handle contention for the AT ports, make it available via a simple call but also be the webserver and possibly an ssh server and not require any other dependencies.  It could handle status parsing as well. Go is compiled so I think the resource requirements would actually be quite small and it should be even faster than linking together lots of scripts. (in theory anyhow)

Also do you know if the RM521 is going to get any firmware updates? I just read something the other day that said that it may not in the long run.. That would kinda suck. I'm almost ready to trade someone my 521 for a 520 since I never need 3CA. I would love to get 71+41+41 working. I can pull it in but if I lock the modem on 71 it just doesn't seem to work quite right.. I can ping, I can wget small websites from the modem but bigger websites seem to not work and basically nothing seems to work from the ethernet side when locked to a specific channel.

**Reply by snowzach** [reply]
I figured it out last night... It's working now.. Got it polling for basic stuff. Will get the server running this weekend. It has a web endpoint that returns the current status in JSON format. The only hard part now really is converting that giant parsing bash script to Go. It's really complicated, chock full of regex and no comments lol.

**Reply by iamromulan** [reply]
Check out the new development-v2 branch
@dr-dolomite has this done client-side now so we don't need the parse script anymore.

**Reply by snowzach** [reply]
Interesting! That could be nice.. Honestly that makes it a lot easier if using this go binary is something y'all want to entertain. It eliminates all the `cat` and `socat` stuff and handles concurrent access to the AT interface in a nice way that you won't have to worry about it. 50 processes could request an AT command all at once and it would handle it and return each one it's response properly. It should require very little overhead, very possibly less than all the different shell scripts and stuff.

I was also thinking you could run this from another machine with the modem connected via USB and it will still work. You just have to start it pointed to the USB interface instead of the local smd one. Part of me likes the option of the modem completely stock and the management interface running in a docker container from another machine managing it from a USB port with very little changes. 

I could also setup a websocket interface so the frontend could schedule commands.. Hey backend, run this status command every X seconds and stream me the results. Food for thought. @dr-dolomite I haven't looked at what you have but if you're interested, I'm happy to do whatever backend work to make talking to the modem easy. It should also easily replace lighttpd. That would basically make opkg completely optional (unless you want it for ttyd, ssh or whatever)

**Reply by iamromulan** [reply]
Not gonna lie, I know nothing about Go. However, as of now (main branch) the 3 separate AT streams are very much needed. 2 of them are needed by simple admin (ttyOUT --> smd11 for the automatic commands to be parsed and ttyOUT2 ---> smd7 for user initiated commands) while 1 was left for use over USB (at_mdm0). I can tell you sending 2 commands at once to the same smd messes it up for sure (had to reboot sometimes to get it back). If we could achieve only needing to use smd11 that would be ideal, however I imagine there would need to be some sort of AT command queue system. Is that something a Go executable could take care of?
Also, i would love a command line tool for windows that can also send AT commands. It is my goal to make this super easy for everyone to deploy (the whole original idea of the toolkit). If i can get a windows AT CLI then maybe I can get the Windows Autoinstaller https://github.com/iamromulan/RM520N-GL to take care of more stuff all at once.

**Comment by snowzach** [comment]
Yeah, so the "quectool" takes care of the command queue bit already. That's working fine. As for a command line tool for Windows, I will look into that. It should be possible as long as I can figure out how to open the USB port in Windows. I do most of my work in Linux.

**Reply by snowzach** [reply]
Alrighty, so I have everything working.. In order for it work it needs to have a daemon running owning the port to ensure there's a queue talking to the port. So what's the best way to interact with it.. 

Right now it runs webserver.. You can use wget on the box to talk to it. Is that sufficient. It would have to listen on another port if you want to keep lighttpd. I can make it do everything lighttpd can do or we can just make it only talk to the port and you could use cgi-bin to talk to the daemon. You can't listen to a dev device and echo to it and then listen to get the result because it won't maintain a session so you get your response and not everyone elses also. 

I'm happy to help make it work how would be the most helpful. Just need to sort out what that is.
